structuredheadings/plugin.js

```diff
@@ -1,67 +1,139 @@
 (function () {
   CKEDITOR.plugins.add("structuredheadings", {
-    icons: "structuredheadings",
+    icons: "autonumberheading, matchheading",
     init: function (editor) {
       // list of elements allowed to be numbered
       var allowedElements = ["h1", "h2", "h3", "h4", "h5", "h6"];
 
       editor.addContentsCss(this.path + "styles/numbering.css");
 
       // Helper function to set button state
-      var setCommandState = function (state) {
+      var setCommandState = function (command, state) {
         switch (state) {
         case "on":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_ON);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_ON);
           break;
         case "off":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_OFF);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_OFF);
           break;
         case "disabled":
-          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_DISABLED);
+          editor.getCommand(command).setState(CKEDITOR.TRISTATE_DISABLED);
           break;
         }
 
       };
 
+      var isNumbered = function (element) {
+        if (element.hasClass("autonumber")) {
+          return true;
+        } else {
+          return false;
+        }
+      };
+
       // The command to add an autonumbering class to selection
       editor.addCommand("autoNumberHeading", {
         allowedContent: "h1(*); h2(*); h3(*); h4(*); h5(*); h6(*)",
-        requiredContent: allowedElements.join(";"),
         startDisabled: true,
         exec: function () {
           var element = editor.getSelection().getStartElement();
 
-          if (!element.hasClass("autonumber")) {
+          if (!isNumbered(element)) {
             element.addClass("autonumber");
-            setCommandState("on");
+            setCommandState("autoNumberHeading", "on");
           } else {
             element.removeClass("autonumber");
-            setCommandState("off");
+            setCommandState("autoNumberHeading", "off");
           }
         }
       });
 
+      editor.addCommand("matchHeading", {
+        startDisabled: true,
+        exec: function () {
+          var element = editor.getSelection().getStartElement();
```

Comment:
I'm not sure how exactly to handle a selection, since the specs for these commands all deal with modifying one singular element, which is why I used getStartElement. If the user has a selection range, the commands are applied to the element where the selection range starts. It treats a selection range as a collapsed selection at the start point. So having a selection range doesn't break anything, and this method supports the use case where the user has selected the full word/sentence of the header rather than just positioning inside of it. (Which is technically a range, but still only inside one element)